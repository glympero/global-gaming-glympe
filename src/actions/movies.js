//SET_MOVIES
const setMovie = (movies) => ({
    type: 'SET_MOVIES',
    movies
  });

// REMOVE_MOVIE
const removeMovie = ({ id } = {}) => ({
    type: 'REMOVE_MOVIE',
    id
});

// EDIT_MOVIE
const editMovie = (id, updates) => ({
    type: 'EDIT_MOVIE',
    id,
    updates
});

export { setMovie, removeMovie, editMovie };