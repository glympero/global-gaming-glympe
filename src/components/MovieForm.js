import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';

class MovieForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      title: props.movie ? props.movie.title : '',
      genre: props.movie ? props.movie.genre : '',
      directorName: props.movie ? props.movie.director.name : '',
      directorGender: props.movie ? props.movie.director.gender : '',
      release_date: props.movie ? moment(props.movie.release_date, 'DD MMMM YYYY') : '',
      error: ''
    }
  }

  onTitleChange = (e) => {
    const title = e.target.value;
    this.setState(() => ({
      title
    }));
  }
  
  onSubmit = (e) => {
    e.preventDefault();
    if(
      !this.state.title || 
      !this.state.genre ||
      !this.state.directorName || 
      !this.state.directorGender || 
      !this.state.release_date
    ) {
      this.setState(() => ({
        error: 'All fields are required'
      }));
    }else {
      this.setState(() => ({
        error: ''
      }));
      const title = this.state.title;
      const genre = this.state.genre;
      const directorName = this.state.directorName;
      const directorGender = this.state.directorGender;
      const release_date = this.state.release_date;
      this.props.onSubmit({ 
        title,
        genre,
        directorName,
        directorGender,
        release_date
      });
    }
  }

  render() {
    return (
      <form className='form' onSubmit={this.onSubmit}>
        { !!this.state.error && <p className='form__error'>{this.state.error}</p>}
        <input className='text-input' type='text' value={this.state.title} onChange={this.onTitleChange} placeholder='title' autoFocus/>
        <input className='text-input' type='text' value={this.state.genre} disabled placeholder='gender' />
        <input className='text-input' type='text' value={this.state.directorName} disabled placeholder='director name' />
        <input className='text-input' type='text' value={this.state.directorGender} disabled placeholder='director gender'/>
        <input className='text-input' type='text' value={this.state.release_date} disabled placeholder='release date' />
        <button className='button'>Edit Title / Close</button>
      </form>
    );
  }
}

MovieForm.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  movie: PropTypes.object.isRequired
};

export default MovieForm;