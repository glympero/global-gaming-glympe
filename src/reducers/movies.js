const moviesReducerDefaultState = [];

const moviesReducer = (state = moviesReducerDefaultState, action) => {
    switch (action.type) {
        case 'SET_MOVIES':
            return action.movies;
        case 'REMOVE_MOVIE':
            return state.filter(movie => movie.id !== action.id);
        case 'EDIT_MOVIE':
            return state.map(movie => {
                if(movie.id === action.id) {
                    return {
                        ...movie,
                        ...action.updates
                    }
                }
                return movie;
            })
        default:
            return state;
    }
};

export default moviesReducer;