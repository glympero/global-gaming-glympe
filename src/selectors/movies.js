const getVisibleMovies = (movies, filters) => {
    const {text, sortBy, genre } = filters;
    return movies.filter(movie => {
        const textMatched = typeof text === 'string' && movie.title.toLowerCase().indexOf(text.toLowerCase()) !== -1;
        const genreMatched = typeof genre === 'string' && genre === 'all' ? true : movie.genre.toLowerCase().indexOf(genre.toLowerCase()) !== -1;
        return textMatched && genreMatched;
    }).sort((a, b) => {
        if(sortBy === 'title') {
            return a.title > b.title ? 1 : -1;
        }else if (sortBy === 'director') {
            return a.director.toLowerCase() > b.director.toLowerCase() ? 1 : -1;
        }
    });
};

export default getVisibleMovies;