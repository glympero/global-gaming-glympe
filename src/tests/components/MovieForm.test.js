import React from 'react';
import { shallow } from 'enzyme';
import MovieForm from '../../components/MovieForm';
import movies from '../../fixtures/data';

let onSubmit, wrapper, match
beforeEach(() => {
    onSubmit = jest.fn();
    match = { params: 1 }

    wrapper = shallow(
        <MovieForm 
            onSubmit={onSubmit} 
            match={match}
            movie={movies[0]}
        />
    )
});

test('should render MovieForm correctly', () => {
    expect(wrapper).toMatchSnapshot();
});