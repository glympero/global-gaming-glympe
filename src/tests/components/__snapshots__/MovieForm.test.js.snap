// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render MovieForm correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="title"
    type="text"
    value="Absolute Bite"
  />
  <input
    className="text-input"
    disabled={true}
    placeholder="gender"
    type="text"
    value="Drama@april.biz"
  />
  <input
    className="text-input"
    disabled={true}
    placeholder="director name"
    type="text"
    value="Walsh Mertie"
  />
  <input
    className="text-input"
    disabled={true}
    placeholder="director gender"
    type="text"
    value="Female"
  />
  <input
    className="text-input"
    disabled={true}
    placeholder="release date"
    type="text"
    value={"2017-06-11T21:00:00.000Z"}
  />
  <button
    className="button"
  >
    Edit Title / Close
  </button>
</form>
`;
